examples/bigquery_user_info_updater_tool/cron/bigquery_user_info_updater/updater_tools/user_info_updater.py

```diff
@@ -0,0 +1,168 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import datetime
+from dateutil import tz
+import logging
+
+from google.cloud import bigquery
+
+INITIAL_TIMESTAMP = datetime.datetime(1900, 1, 1, 00, 00, 00, 000,
+                                      tzinfo=tz.tzutc())
+
+
+class UserInfoUpdater(object):
+    """Class for updating Identity data in BigQuery.
+
+    Holds methods for gathering updates to identity data and merging
+    updates into the master identity table.
+
+    Attributes:
+        bq_client(google.cloud.bigquery.client.Client): Client to hold
+            configurations needed for BigQuery API requests.
+        project_id(str): Id of project holding BQ resources.
+        dataset_id(str): Id of dataset holding Identity BQ tables.
+        updates_table_id(str): Id of table holding updates for the identity
+            table.
+        temp_updates_table_id(str): Id of table that deduped update rows will
+            be written to.
+        final_table_id(str): Id of identity table holding golden record for
+            each user.
+
+    """
+
+    def __init__(
+            self,
+            project_id,
+            dataset_id,
+            updates_table_id,
+            temp_updates_table_id,
+            final_table_id
+    ):
+        self.bq_client = bigquery.Client(project=project_id)
+        self.project_id = project_id
+        self.dataset_id = dataset_id
+        self.dataset_ref = self.bq_client.dataset(self.dataset_id)
+        self.updates_table_id = updates_table_id
+        self.temp_updates_table_id = temp_updates_table_id
+        self.final_table_id = final_table_id
+
+    def get_max_ingest_timestamp(self):
+        """Gets the max timestamp that was set during the latest merge.
+
+        Returns:
+            latest_merge as a timestamp in string format.
+        """
+        get_last_max_ts_config = bigquery.QueryJobConfig()
+        get_last_max_ts_config.use_legacy_sql = False
+        get_last_max_timestamp_query = self.bq_client.query(
+            query='SELECT max(ingestTimestamp) as max_ingest_timestamp '
+                  'FROM `{0:s}.{1:s}.{2:s}`'.format(
+                    self.project_id,
+                    self.dataset_id,
+                    self.temp_updates_table_id
+                  ),
+            job_config=get_last_max_ts_config,
+            location='US'
+        )
+        get_last_max_timestamp_query.result()
+        results = list(get_last_max_timestamp_query)
+        max_ingest_timestamp = results[0]['max_ingest_timestamp']
+        if not max_ingest_timestamp:
+            max_ingest_timestamp = INITIAL_TIMESTAMP
+
+        return max_ingest_timestamp.strftime('%Y-%m-%d %H:%M:%S.%f %Z')
```

Comment:
jaketf on Aug 27, 2019 Member
Why are we casting as str?
What are we casting from?
If datetime we should use strftime